src/java/org/apache/cassandra/db/virtual/CollectionVirtualTableAdapter.java

```diff
@@ -0,0 +1,589 @@
+                    case CLUSTERING:
+                        fiterable.get(type)[cIdx++] = value.get();
+                        break;
+                    case REGULAR:
+                    {
+                        if (columnFilter.equals(ColumnFilter.NONE))
+                            break;
+
+                        // Push down the column filter to the walker, so we don't have to process the value if it's not queried
+                        ColumnMetadata cm = columnMetas.computeIfAbsent(columnName, name -> metadata.getColumn(
+                            ByteBufferUtil.bytes(name)));
+                        if (columnFilter.queriedColumns().contains(cm))
+                            cells.put(cm, value);
+
+                        break;
+                    }
+                    default:
+                        throw new IllegalStateException("Unknown column type: " + type);
+                }
+            }
+        });
+
+        return new CollectionRow(() -> makeRowKey(metadata, fiterable.get(Column.Type.PARTITION_KEY)),
+                                 makeRowClustering(metadata, fiterable.get(Column.Type.CLUSTERING)),
+                                 clustering ->
+                                 {
+                                     Row.Builder rowBuilder = BTreeRow.unsortedBuilder();
+                                     rowBuilder.newRow(clustering);
+                                     cells.forEach((column, value) -> {
+                                         Object valueObj = value.get();
+                                         if (valueObj == null)
+                                             return;
+                                         rowBuilder.addCell(BufferCell.live(column, NO_DELETION_TIME,
+                                                                            decompose(column.type, valueObj)));
+                                     });
+                                     return rowBuilder.build();
+                                 });
+    }
+
+    private static class CollectionRow
+    {
+        private final Supplier<DecoratedKey> key;
+        private final Clustering<?> clustering;
+        private final Supplier<Row> rowSup;
+
+        public CollectionRow(Supplier<DecoratedKey> key, Clustering<?> clustering, Function<Clustering<?>, Row> rowSup)
+        {
+            this.key = new ValueHolder<>(key);
+            this.clustering = clustering;
+            this.rowSup = new ValueHolder<>(() -> rowSup.apply(clustering));
+        }
+
+        private static class ValueHolder<T> implements Supplier<T>
+        {
+            private final Supplier<T> delegate;
+            private volatile T value;
+
+            public ValueHolder(Supplier<T> delegate)
+            {
+                this.delegate = delegate;
+            }
+
+            @Override
+            public T get()
+            {
+                return value == null ? value = delegate.get() : value;
```

Comment:
btw IDEA is not happy about this, we are assigning and returning, it tells to split this 

    if (value == null)
        value = delegate.get()

    return value

Are we doing this because `value` is `volatile`?

Reply:
Fixed.